icu/icu4c/source/test/intltest/uprefstest.cpp

```diff
@@ -0,0 +1,437 @@
+
+        if (language.length() + 1 > cchData)
+        {
+            *status = U_BUFFER_OVERFLOW_ERROR;
+            return -1;
+        }
+        wcsncpy(lpLCData, language.c_str(), cchData);
+        *status = U_ZERO_ERROR;
+        return language.length();
+
+    case LOCALE_ICALENDARTYPE | LOCALE_RETURN_NUMBER:
+        if (cchData == 0)
+        {
+            *status = U_ZERO_ERROR;
+            return 2;
+        }
+        if (cchData < 2)
+        {
+            *status = U_BUFFER_OVERFLOW_ERROR;
+            return -1;
+        }
+        *(reinterpret_cast<int32_t*>(lpLCData)) = calendar;
+        *status = U_ZERO_ERROR;
+        return 2;
+
+    case LOCALE_SINTLSYMBOL:
+        if (cchData == 0)
+        {
+            *status = U_ZERO_ERROR;
+            return currency.length() + 1;
+        }
+        if (currency.length() + 1 > cchData)
+        {
+            *status = U_BUFFER_OVERFLOW_ERROR;
+            return -1;
+        }
+        wcsncpy(lpLCData, currency.c_str(), cchData);
+        *status = U_ZERO_ERROR;
+        return currency.length();
+
+    case LOCALE_IFIRSTDAYOFWEEK | LOCALE_RETURN_NUMBER:
+        if (cchData == 0)
+        {
+            *status = U_ZERO_ERROR;
+            return 2;
+        }
+        if (cchData < 2)
+        {
+            *status = U_BUFFER_OVERFLOW_ERROR;
+            return -1;
+        }
+
+        *(reinterpret_cast<int32_t*>(lpLCData)) = firstday;
+        *status = U_ZERO_ERROR;
+        return 2;
+
+    case LOCALE_STIMEFORMAT:
+        if (cchData == 0)
+        {
+            *status = U_ZERO_ERROR;
+            return hourCycle.length() + 1;
+        }
+        
+        if (hourCycle.length() + 1 > cchData)
+        {
+            *status = U_BUFFER_OVERFLOW_ERROR;
+            return -1;
+        }
+        wcsncpy(lpLCData, hourCycle.c_str(), cchData);
+        *status = U_ZERO_ERROR;
+        return 0;
+    
+    case LOCALE_IMEASURE | LOCALE_RETURN_NUMBER:
+        if (cchData == 0)
+        {
+            *status = U_ZERO_ERROR;
+            return 2;
+        }
+
+        if (cchData < 2)
+        {
+            *status = U_BUFFER_OVERFLOW_ERROR;
+            return -1;
+        }
+        *(reinterpret_cast<int32_t*>(lpLCData)) = measureSystem;
+        *status = U_ZERO_ERROR;
+        return 2;
+
+    default:
+        *status = U_INTERNAL_PROGRAM_ERROR;
+        return -1;
+    }
+}
+
+#include "uprefs.cpp"
+
+void UPrefsTest::TestGetBCP47Tag1()
+{
+    bool correct = false;
+    char languageBuffer[ARRAY_SIZE] = {0};
+    language = L"en-US";
+    currency = L"USD";
+    hourCycle = L"HH:mm:ss";
+    firstday = 0;
+    measureSystem = 1;
+    calendar = CAL_GREGORIAN;
+    UErrorCode status = U_ZERO_ERROR;
+    std::string expectedValue = "en-US-u-ca-gregory-cu-usd-fw-mon-hc-h23-ms-ussystem\0";
+    
+    if ( uprefs_getBCP47Tag(languageBuffer, ARRAY_SIZE, &status) != 52)
+    {
+        errln("Expected length to be 52, but got: %d\n",strlen(languageBuffer));
```

Comment:
```suggestion
        errln("Expected length to be 52, but got: %d\n", uprv_strlen(languageBuffer));
```
[and all the other places in the file too... :) ]